src/commissaire_service/clusterexec/__init__.py

```diff
@@ -0,0 +1,305 @@
+                        host.address))
+                continue  # Move on to the next one
+            oscmd = get_oscmd(host.os)
+
+            # os_command is only used for logging
+            os_command = getattr(oscmd, command_name)(*command_args)
+            self.logger.info('Executing {} on {}...'.format(
+                os_command, host.address))
+
+            model_instance.in_process.append(host.address)
+            try:
+                params = {
+                    'model_type_name': model_instance.__class__.__name__,
+                    'model_json_data': model_instance.to_json()
+                }
+                self.request('storage.save', params=params)
+            except Exception as error:
+                self.logger.error(
+                    'Unable to save in_process state for "{}" clusterexec '
+                    'due to {}: {}'.format(cluster_name, type(error), error))
+                raise error
+
+            with TemporarySSHKey(host, self.logger) as key:
+                try:
+                    transport = ansibleapi.Transport(host.remote_user)
+                    method = getattr(transport, command_name)
+                    method(host.address, key.path, oscmd, command_args)
+                except Exception as error:
+                    # If there was a failure, set the end_status and break.
+                    end_status = 'failed'
+                    self.logger.error(
+                        'Clusterexec {} for {} failed: {}: {}'.format(
+                            command_args[0], host.address, type(error), error))
+                    break
+
+            # Set the finished hosts.
+            finished_hosts = getattr(model_instance, finished_hosts_key)
+            finished_hosts.append(host.address)
+            try:
+                index = model_instance.in_process.index(host.address)
+                model_instance.in_process.pop(index)
+            except ValueError:
+                self.logger.warn(
+                    'Host {} was not in_process for {} {}'.format(
+                        host.address, command_name, cluster_name))
+            try:
+                params = {
+                    'model_type_name': model_instance.__class__.__name__,
+                    'model_json_data': model_instance.to_json()
+                }
+                self.request('storage.save', params=params)
+            except Exception as error:
+                self.logger.error(
+                    'Unable to save cluster state for "{}" clusterexec '
+                    'due to {}: {}'.format(cluster_name, type(error), error))
+                raise error
+
+            self.logger.info(
+                'Finished executing {} for {} in {}'.format(
+                    command_name, host.address, cluster_name))
+
+        # Final set of command result.
+
+        model_instance.finished_at = datetime.datetime.utcnow().isoformat()
```

Comment:
future thought: We use ``datetime.datetime.utcnow().isoformat()`` often. It might be worth making a function in ``commissaire.util`` that returns ``now`` in the proper format in the event we later need to change.